### DISCUSSION TITLE
Using DOM interactions (useClick, useDismiss, etc.) when the reference element is passed as a prop?

### DISCUSSION THREAD

**Discussion by joeattardi** [floating-ui/floating-ui#1834] [React] [answered]
I have a Menu component, one of its props is a ref that is expected to be set on the menu trigger button, also the reference element.

I am able to get the floating piece accomplished by manually setting the reference ref:

```
const Menu({ triggerRef }) => {
  const { refs, x, y, floating, reference, strategy, context } = useFloating(...);

  refs.reference.current = triggerRef.current;
});
```

So my floating element is positioned over the button pointed to by the ref. But I'm not sure how to use DOM interactions here. Since `useInteractions` returns the `getReferenceProps` function that would normally be spread onto the reference element's props. But if all I have is a ref to a button passed in from elsewhere, how do I get these props attached?

Thanks!

**Comment by atomiks** [comment, marked as answer]
You would need to hoist the hooks above the Menu component to where the reference element JSX is, then pass the data down to the Menu component as props.

It's much better if you use `children` prop composition though, so then you attach reference props via `cloneElement` as the CodeSandbox examples show. The "external" reference element pattern tends to have poor ergonomics for reusable components.

This line also doesn't seem right as you can't write or read refs during render:

```js
refs.reference.current = triggerRef.current;
```

The best two options are either `children` prop or `render` prop:

```js
<Popover>
  <Button>This is the reference. Popover can attach the props to me.</Button>
</Popover>
```

```js
<Popover renderAnchor={(props) => <Button {...props}>I am spreading the props onto me.</Button>}>
  This is the popover content.
</Popover>
```

**Reply by joeattardi** [reply]
This is very helpful, thanks! And thanks for an amazing library.
